File: Programmers/BoHwa/2주차(Brute_Force)/소수찾기.java
import java.util.ArrayList;

class Solution {
    ArrayList<Integer> arr = new ArrayList<Integer>(); //만들어진 숫자를 저장하는 List
    boolean visited[];//방문했는지 확인하는 배열

    public int solution(String numbers) {
        int answer = 0;
        String tmp = "";
        visited = new boolean[numbers.length()];

        for (int i = 1; i <= numbers.length(); i++) { //i 자릿수 만듦
            makeNum(numbers, tmp, i);
        }

        for (int x : arr) {
            if (isPrime(x)) answer++;
        }

        return answer;
    }

    public void makeNum(String num, String tmp, int len) {
        if (tmp.length() == len) {//현재 문자열의 길이가 원하는 자릿수와 같다면
            if (!arr.contains(Integer.parseInt(tmp))) //현재 문자열을 int로 전환하여 ArrayList에 포함되어 있는지 확인
                arr.add(Integer.parseInt(tmp)); //없으면 추가
            return; //돌아감
        }
        for (int i = 0; i < num.length(); i++) { //문자열 길이만큼 반복
            if (visited[i]) continue; //방문했던 숫자라면 넘어감
            tmp += num.charAt(i);
            visited[i] = true;
            makeNum(num, tmp, len);
            visited[i] = false;
            tmp = tmp.substring(0, tmp.length() - 1); //마지막 자리 숫자 제외 후 갱신
        }

    }

    public boolean isPrime(int n) {
        if (n == 0 || n == 1) return false;

        for (int i = 2; i <= Math.sqrt(n); i++) { //테스트 케이스 2, 10, 11 통과, =을 포함시켜야 한다.
            if (n % i == 0) return false;
        }
        return true;
    }

}

File: Programmers/Segun/2주차(Brute_force)/소수 찾기.cpp
//흩어진 종이조각을 '조합'으로 맞춰주고 그 만들어진수가 소수인지를 확인한 뒤에 result++ 해준다.

#include <string>
#include <vector>

using namespace std;

bool visited[11];   //방문 여부 변수
char arr[11];   //뽑은 종이 조각을 저장
int result; //결과값
bool chk[11111111]; //이미 확인했던 값인지 확인

bool decimal(int x) {   //소수를 판별해주는 함수
    if (chk[x] == false) { chk[x] = true; } //처음들어오는 값이라면 확인처리
    else { return false; }  //이미 들어왔던 수라면 false 리턴
    for (int i = 2; i <= x / 2; i++) {  //x가 1과 자기자신외에 다른 수로 나누어지는지 확인
        if (x % i == 0) { return false; }   //나누어진다면 소수가 아님
    }
    
    return true;    //통과했다면 소수
}

//흩어진 종이조각을 맞춰주는 함수
void DFS(int cnt, string num) {
    if (cnt > 0 && cnt <= num.size()) { //수가 1개이상 주어진 수의 길이 이하라면 소수 판별해보기
        string str;
        for (int i = 0; i < cnt; i++) { str.push_back(arr[i]); }    
        //string을 바로 접근하면 오류발생해서 char 배열 사용
        if (decimal(stoi(str)) == true) { result++; } } //소수 확인
    for (int i = 0; i < num.size(); i++) {  //주어진 문자열의 길이만큼 탐색
        if (visited[i] == false) {  //이미 탐색했던곳인지 확인
            visited[i] = true;  //아니라면 방문 표시
            arr[cnt] = num[i];  //고른 종이 조각을 arr에 저장
            DFS(cnt + 1,num);   //다음 종이 조각 뽑기
            visited[i] = false; //종이 조각 뽑고 확인까지 끝낫다면 다시 방문표시를 없애줌
        }
    }
}

int solution(string numbers) {
    chk[0] = true; chk[1] = true;   //0,1은 소수가 아님
    DFS(0, numbers);
    return result;
}
